// _vcs/esova/esova.js
// const { h, hsvg } = require('_h')
var h = require('virtual-dom/h')
const s           = require('_s')

const header      = require('_vcs/header')
// const skilltree   = require('_vcs/skilltree')
// const agent       = require('_vcs/agent')

var path          = require('path')
var urify         = require('urify')

module.exports    = RootComponent

function RootComponent (db, engine) {

  // @TODO: where to set boilerplate in general?
  document.title = 'Esova'  // @TODO: set in other ways - from package.json?

  // html {
  //   box-sizing: border-box;
  // }
  // *, *:before, *:after {
  //   box-sizing: inherit;
  // }
  var x = s(`
    body {
      display: flex;
      margin: 0;
      flex-direction: row;
      align-items: stretch;
      background-color: ${'#526E96'};
    }
    esova {
      display: flex;
      flex-direction: column;
      width: 99%;
      height: 100%;
    }
    header {
      position: relative;
      left: 8;
      min-height: 165px;
    }
    content {
      display: flex;
      flex-direction: row;
      justify-content: center;
      align-items: stretch;
      flex-grow: 1;
    }
    footer {
      min-height: 165px;
    }
  `)
  document.body.className = x('body')
  return function render () {
    var vtree = h('div', { className: x('esova') }, [
      h('div', { className: x('header') }, header(db, engine)),
      h('div', { className: x('content') }, content(db, engine)),
      h('div', { className: x('footer') }, footer(db, engine))
    ])
    return vtree
  }
}
function footer () { return '' }

var forward = urify(path.join(__dirname, 'forward.png'))
var forward_pressed = urify(path.join(__dirname, 'forward-pressed.png'))
var forward_arrow = urify(path.join(__dirname, 'forward-arrow.png'))
var forward_pressed = urify(path.join(__dirname, 'forward-pressed.png'))
var back = urify(path.join(__dirname, 'back.png'))
var back_pressed = urify(path.join(__dirname, 'back-pressed.png'))
var back_arrow = urify(path.join(__dirname, 'back-arrow.png'))
var back_pressed = urify(path.join(__dirname, 'back-pressed.png'))

function content (db, engine) {
  var x = s(`
    skillbox {
      display: flex;
      flex-direction: column;
      justify-content: center;
      align-items: center;
      min-height: 100px;
      width: 50%;
    }
    skillbox-nav {
      display: flex;
      flex-direction: row;
      align-items: center;
      justify-content: center;
      min-height: 40px;
      width: 100%;
      margin-bottom: 20px;
    }
    skillbox-description {
      display: flex;
      align-items: center;
      justify-content: center;
      width: 100%;
      height: 100%;
      background-color: ${'#fff'};
      border-top-right-radius: 10px;
      border-bottom-right-radius: 10px;
      border-top-left-radius: 10px;
      border-bottom-left-radius: 10px;
    }
    navright {
      padding: 10px;
      text-align:center;
      background-image: url(${forward});
      background-position: center;
      background-size: 100%;
      background-repeat: no-repeat;
    }
    navleft {
      padding: 10px;
      text-align:center;
      margin-right: 100px;
      background-position: center;
      background-image: url(${back});
      background-size: 100%;
      background-repeat: no-repeat;
    }
  `)
  return h('div', { className: x('skillbox') }, [
    h('div', { className: x('skillbox-nav') }, [
      h('div', { className: x('navleft') }, [
        h('img', { src: back_arrow })
      ]),
      h('div', { className: x('navright') }, [
        h('img', { src: forward_arrow })
      ]),
    ]),
    h('div', { className: x('skillbox-description') }, 'lorem ipsum...')
  ])
}
