// _vcs-promocode/promocode.js
'use strict'
/******************************************************************************
  STACK BASE
******************************************************************************/
var { h, hsvg } = require('_h')
var s           = require('_s')
/******************************************************************************
  STACK EXTEND
******************************************************************************/
var path        = require('path')
var urify       = require('urify')
var minixhr     = require('minixhr')
var defined     = require('defined')
/******************************************************************************
  CUSTOM
******************************************************************************/
var auth = require('_auth')
/******************************************************************************
  ASSETS
******************************************************************************/
var start = urify(path.join(__dirname, 'start.png'))
var header = urify(path.join(__dirname, 'header.png'))
var closebutton = urify(path.join(__dirname, 'closebutton.png'))
/******************************************************************************
  MAIN
******************************************************************************/
module.exports = PromocodeComponent

function PromocodeComponent (db, engine) { // SIGNUP COMPONENT - OVERLAY


  /*****************************/
  // THEME + LOCAL CACHE + DEFAULT VALUES
  var esovablue = '#526E96'
  var esovalightblue = '#edf2fa'
  var esovalightgreen = '#35E38A'
  var esovared = 'FA9B68'
  var FREE_LESSONS  = 4
  var STATUS = {
    // message: 'ok',
    // message: 'hide',
    // message: 'wrong promocode',
    message: 'not ready'
    // message: 'unmet requirements'
    // promocode: ''
  }


  /*****************************/
  // INITIALIZE STYLING
  var x = s(`
    promocode {
      display: flex;
      flex-direction: column;
      height: 100%;
      width: 100%;
      justify-content: flex-start;
      align-items: center;
      z-index: 10;
      background-color: hsla(0, 0%, 0%, 0.19);
      border-radius: 20px;
    }
    welcome {
      margin-top: 245px;
      display: flex;
      box-sizing: content-box;
      flex-direction: column;
      justify-content: flex-start;
      align-items: center;
      min-width: 200px;
      width: 80%;
      max-width: 475px;
      min-height: 280px;
      height: 27%;
      padding: 0 10px 80px 10px;
      background-color: hsla(0, 0%, 100%, 1);
      color: black;
      font-family: arial;
      border-radius: 15px;
      font-size: 3vmin;
    }
    message {
      color: ${esovalightgreen};
      font-size: 0.8em;
      text-align: center;
      font-weight: normal;
      font-family: Avenir Roman;
      margin-top: 0px;
      padding: 40px 105px 10px 105px;
    }
    codeinput {
      padding-top: 55px;
    }
    errormessage {
      color: ${esovared};
      font-size: 0.7em;
      text-align: center;
      font-weight: normal;
      font-family: Avenir Roman;
      margin-top: 0px;
      padding: 20px 105px 7px 105px;
    }
    promoinput {
      background-color: ${esovalightblue};
      font-size: 2vmin;
      padding: 10px;
      border: 1px solid ${esovablue};
      border-top-left-radius: 15px;
      border-bottom-left-radius: 15px;
      color: white;
      border-right: 0;
      font-family: arial;
      text-align: center;
    }
    signup {
      background-color: ${esovablue};
      border: 1px solid ${esovablue};
      border-top-right-radius: 15px;
      border-bottom-right-radius: 15px;
      padding: 10px;
      font-family: arial;
      color: white;
      font-size: 2vmin;
      cursor: pointer;
    }
    notready {
      display: flex;
      align-items: center;
      justify-content: center;
    }
    startnow {
      background-color: ${esovablue};
      border: 1px solid ${esovablue};
      border-radius: 15px;
      padding: 10px;
      font-family: arial;
      color: white;
      font-size: 2vmin;
      cursor: pointer;
    }
    closebutton {
      align-self: flex-end;
      padding-top: 10px;
    }
  `)


  /*****************************/
  // CHILD COMPONENTS & USED MODULES
  // ...
  function returnNotReady () {
    return h('div', [
      // @TODO @NOTE: add to translation file
      h('h2', { className: x('message') }, 'Kreiranje eSova naloga je moguće na kraju četvrte lekcije.'),
      h('div',  { className: x('notready') }, [
        // @TODO @NOTE: add to translation file
        h('button', { className: x('startnow') }, 'Počnite sa lekcijama')
      ])
    ])
  }

  function returnWrongPromocode () {
    return h('div', [
      // @TODO @NOTE: add to translation file
      h('h2', { className: x('errormessage') }, 'Uneli ste pogrešan kod, pokušajte ponovo.')
    ])
  }


  /*****************************/
  // RENDER COMPONENT
  function render () {
    // @TODO @NOTE: when at lesson 4 end for signin => set currentLesson to 5
    // => notifiy to finish first 4 lessons first
    // => ask promo
debugger
    var vtree = STATUS.message === 'hide' ?
      '' // renders nothing to hide the overlay
    : h('div', { className: x('promocode') }, [
      h('div', { className: x('welcome') }, [
        h('img', { src: closebutton, className: x('closebutton')}),
        h('img', { src: header, className: x('header-picture')}),
        STATUS.message === 'not ready' ?
        returnNotReady()
        : h('div', { className: x('codeinput') }, [
          h('input', {
            className: x('promoinput'),
            name: 'promocode',
            onchange: function (event) {
              STATUS.promocode = event.target.value
            },
            onkeyup: function (event) {
              var key = event.which
              if (key === 13) { // enter
                db.put('esova-promocode', STATUS.promocode, function (error) {
                  // auth === ACTION_signIn(...)
                  // param is/was set via db.put ...
                  auth(db, function (error, help) { })
                })
              }
            },
            // @TODO @NOTE: add to translation file
            placeholder: 'Upišite svoju kodu'
          }),
          h('button', {
            className: x('signup'),
            onclick: function () {
              // @IDEA
              // The "SIGNAL" is like a "BEHAVIOR" instead of "EVENT"
              // => when it changes, listeners are informed
              // => the signal (=value of the "event") stays the same until
              //    its changed to something else
              db.put('esova-promocode', STATUS.promocode, function (error) {
                // auth === ACTION_signIn(...)
                // param is/was set via db.put ...
                auth(db, function (error, help) { })
              })
            },
          }, 'Potvrdi')
        ]),
        STATUS.message === 'wrong promocode' ?
        returnWrongPromocode()
        : ''
      ])
    ])
    return vtree
  }


  /*****************************/
  // INCOMING COMMUNICATION CHANNELS
  db.on('put', function (key, value) {
    value = defined(value, {})
    if (key === 'OVERLAY') {
      debugger
      value = JSON.parse(value)
      db.get('esova-currentLesson', function (error, lastlesson) {
        if (value.status === 'no valid promocode') {
          var deny = lastlesson ? lastlesson <= FREE_LESSONS : false
          if (!deny) engine.push({ // AUTH FAIL + currentLesson <= FREE_LESSONS
            type: 'vtree',
            key: 'promocode',
            value: render('allow')
          })
          else if (value.status === 'allow invalid') engine.push({
            type: 'vtree',
            key: 'promocode',
            value: '1+1'
          })
          else if (value.status === 'clear') engine.push({
            type: 'vtree',
            key: 'promocode',
            value: ''
          })
          else engine.push({ // AUTH FAIL + currentLesson 5+
            type: 'vtree',
            key: 'promocode',
            value: render('deny')
          })
        } else if (value.status === 'unmet requirements') {
          engine.push({
            type: 'vtree',
            key: 'promocode',
            value: render('allow')
          })
        }
        else {
          // ... // console.error('STATUS '+value+' is unknown')
        }
      })
    }
    /********************************/
  })


  /*****************************/
  // INITIALIZE
  console.log('--------------------------------')
  console.log('INITIALIZE: promocode() component')
  console.error('doesnt work yet actually!, right???')
  console.log('--------------------------------')
  // components at least do not return vtree ...but maybe some kind of API
  // ... to set "default wireup" in context of users instantiating "usage code"
  return undefined // components don't return anything! @TODO: remove this line


  /*****************************/
  // HELPER
  // ...


}
