// _vcs/header/header.js
// const { h, hsvg } = require('_h')
var h             = require('virtual-dom/h')
const s           = require('_s')

var path          = require('path')
var urify         = require('urify')


var owl = urify(path.join(__dirname, 'owl.png'))
var owlbase = urify(path.join(__dirname, 'owlbase.png'))
var blocks = urify(path.join(__dirname, 'blocks.png'))

module.exports    = header

function header (db, engine) {

  var x = s(`
    header {
      display: flex;
      margin-top: 40px;
      align-items: center;
      justify-content: center;
    }
    menubar {
      box-sizing: border-box;
      position: relative;
      display: flex;
      border-radius: 50px;
      height: 85px;
      justify-content: space-between;
      align-items: center;
      background-color: ${'#fff'};
      min-width: 550px;
      width: 90%;
      max-width: 950px;
    }
    logo {
      position: relative;
      top: 2%;
      display: flex;
      justify-content: center;
      align-items: center;
      width: 150px;
      height: 150px;
      padding-left: 55px;
      padding-right: 60px;
    }
    logo-owl {
      z-index: 3;
    }
    logo-owlbase {
      position: absolute;
      top: 1%;
      left: 21.5%;
      z-index: 1;
    }
    logo-blocks {
      position: absolute;
      top: 5%;
      left: 8%;
      height: 130px;
      width: 235px;
      z-index: 2;
    }

    usernav {
      display: flex;
      justify-content: space-between;
      align-items: center;
      height: 85px;
      flex-grow:1;
      border-top-right-radius: 50px;
      border-bottom-right-radius: 50px;
      border-left: 2px solid ${'#32A971'};
      background-color: #35E38A;
    }
    esovanav {
      display: flex;
      justify-content: space-between;
      align-items: center;
      height: 85px;
      flex-grow:1;
      border-top-left-radius: 50px;
      border-bottom-left-radius: 50px;
    }
  `)
  var vtree = h('div', { className: x('header') }, [
    h('div', { className: x('menubar') }, [
      h('div',{ className: x('esovanav') }, [
        h('a', { className: x('home'), href: '' }, "HOME"),
        h('a', { className: x('chat'), href: '#gitter' },"CHAT")
      ]),
      h('div', { className: x('logo') }, [
        h('img', { className: x('logo-owl'), src: owl }),
        h('img', { className: x('logo-owlbase'), src: owlbase }),
        h('img', { className: x('logo-blocks'), src: blocks })
      ]),
      h('div', { className: x('usernav') }, [
        h('a', { className: x('auth'), href: '#auth' },"AUTH"),
        h('a', { className: x('profile'), href: '#profile' },"PROFIL")
      ])
    ])
  ])
  return vtree
}
