// _auth/auth.js
'use strict'
/****************************************************************************
  STACK
****************************************************************************/
const WINDOW  = require("global/window")
const minixhr = require('minixhr')
const stream  = require('readable-stream')
/******************************************************************************
  CUSTOM
******************************************************************************/
const NO_ERROR    = null
const NO_ENDPOINT = new Error('no endpoint given')
const NO_CLIENTID = new Error('no clientID given')
const NO_RIGHTS   = new Error('no right requests given')
/******************************************************************************
  MAIN
******************************************************************************/
module.exports = GithubAuth
function GithubAuth (db, cb) {
  var settings  = {  // DIGITAL OCEAN BACKEND
    endpoint        : 'http://45.55.222.16/github_auths/access_token',
    clientID        : '36645fe5dcc1d41309ab',
    rights          : ['gist']
  }
  db.get('esova-token', function (error, token) {
    error ? auth(settings, function start (error, token) {
      if (error) { throw error }
      db.put('esova-token', token)
      cb(token)
    }) : cb(token)
  })
}
function auth (x, cb) {
  if (WINDOW.opener) {
    WINDOW.opener.sendCode(location.search.split('=')[1])
    WINDOW.close()
  } else {
    if (!x.endpoint) { return cb(NO_ENDPOINT) }
    if (!x.clientID) { return cb(NO_CLIENTID) }
    if (!x.rights)   { return cb(NO_RIGHTS)   }
    WINDOW.sendCode = function sendCode (code) {
      var promocode = db.get('esova-promocode', function (error, promocode) {
        var params = '?code='+code+ (promocode ? '&promocode='+promocode : '')
        minixhr(x.endpoint+params, function (data, _, __, header) {
          // data = { validPromo: false/true, token: null/<token> }
          try { data = JSON.parse(data) }
          catch (e) { cb(e) }
          if (data.validPromo) {
            var token = data.token && data.token.length ? data.token : null
            if (token && token.length < 100) cb(NO_ERROR, data)
            else cb(NO_ERROR, null)
          } else cb(new Error('no valid promo code'))
        })
      })
    }
    var host    = 'http://github.com/login/oauth/authorize?'
    var params  = 'client_id='+x.clientID+'&scope='+x.rights.join('%2C')
    WINDOW.open(host + params, '_blank')
  }
}
