// _githubapi/githubapi.js
'use strict'
var minixhr = require('minixhr')
/************************************************
  CUSTOM
************************************************/
var STATUS = {
  200: success,
  304: unchanged,
  401: unauthorized,
  403: notAvailable,
  404: notFound
}
var host = 'https://api.github.com/'
var param = '?access_token='
var ACTIONS = {
  'user': true
}
/************************************************
  MAIN
************************************************/
module.exports = githubapi

function githubapi (db) {
  return api
  function api (action, cb) {
    if(!ACTIONS[action]) cb(new Error('no action given'))
    else db.get('esova-token', function (error, token) {
      db.get('gAction/'+action, function (error, oldData) {
        oldData = oldData ? JSON.parse(oldData) : {}
        var request = token ?
          buildAuthedRequest(action, token, oldData)
          : buildUnauthedRequest(action, oldData)
        var handler = makeHandler(db, action, oldData, cb)
        minixhr(request, handler)
      })
    })
  }
}
/************************************************
  HANDLER - 200: success
************************************************/
function success (_, db, __, action, ____, newData, header, cb) {
  newData   = JSON.parse(newData)
  var value = JSON.stringify({
    ETag      : header.ETag,
    newData   : newData,
    action    : action
  })
  db.put('gAction/'+action, value, function (error) {
    if (error) cb(error, null)
    else cb(null, newData)
  })
}
/************************************************
  HANDLER - 304: unchanged
************************************************/
function unchanged (_, db, __, ___, oldData, ____, _____, cb) {
  cb(null, oldData)
}
/************************************************
  HANDLER - 401: unauthorized
************************************************/
// @TODO: Just make sure to catch 401 Unauthorized error on requests
// in case the token is no longer valid and ask the user to log in again.
function unauthorized (error, db, status, _, __, ___, ____, callback) {
  var e = error || new Error('status: '+status+' (auth token invalid)')
  e.status = status
  callback(e)
}
/************************************************
  HANDLER - 404: notFound
************************************************/
function notFound (error, db, status, _, __, ___, ____, callback) {
  var e = error || new Error('status: '+status+' (not found)')
  e.status = status
  callback(e)
}
/************************************************
  HANDLER - 403: notAvailable
************************************************/
function notAvailable (error, db, status, _, __, ___, ____, callback) {
  var e = error || new Error('status: '+status+' (not available)')
  e.msg = 'Service temporarily not available - try again in one hour'
  e.status = status
  callback(e)
}
/************************************************
  HANDLER - ???/500: internalError
************************************************/
function internalError (error, db, status, _, __, ___, ____, callback) {
  var e = error || new Error('status: '+status||500+' (Internal Error)')
  e.status = status
  callback(e)
}
/************************************************
  printRemaining
************************************************/
function printRemaining (header) {
  var left  = +header["X-RateLimit-Remaining"]
  var total = +header["X-RateLimit-Limit"]
  var msg   = 'Remaining Requests: ' + left + '/' + total
  console.error(msg)
}
/************************************************
  buildAuthedRequest
************************************************/
function buildAuthedRequest (action, token, oldData) {
  var eTag = oldData.ETag
  return {
    url: host + action + param + token,
    timeout: eTag ? '9999' : null,
    header: {
      "If-None-Match"    : eTag || ''
      // 'X-Requested-With' :'XMLHttpRequest',
      // authorization: 'token ' + token,
      // 'Content-Type': 'application/x-www-form-urlencoded'
      // 'Content-Type': 'application/json'
    }
  }
}
/************************************************
  buildUnauthedRequest
************************************************/
function buildUnauthedRequest (action, oldData) {
  throw new Error('@TODO: not implemented yet!')
  // var eTag = oldData.ETag
  // return {
  //   url: host + action + param + token,
  //   timeout: eTag ? '9999' : null,
  //   header: {
  //     "If-None-Match"    : eTag || ''
  //     // 'X-Requested-With' :'XMLHttpRequest',
  //     // authorization: 'token ' + token,
  //     // 'Content-Type': 'application/x-www-form-urlencoded'
  //     // 'Content-Type': 'application/json'
  //   }
  // }
}
/************************************************
  makeHandler
************************************************/
function makeHandler (db, action, oldData, callback) {
  return function handler (error, newData, _, xhr, header) {
    var handle = STATUS[xhr.status]
    if (!handle) handle = internalError
    printRemaining(header)
    handle(error, db, xhr.status, action, oldData, newData, header, callback)
  }
}
