// _vcs/404/404.js
var h = require('virtual-dom/h')

module.exports = NotFoundComponent

function NotFoundComponent (db, engine) {
  var style1 = {
    display: 'flex',
    flexDirection: 'column',
    color: 'white',
    fontFamily: 'arial',
    width: '100%',
    fontSize: '1.5vmin',
    alignItems: 'center',
    justifyContent: 'flex-start',
  }
  var style2 = {
    fontSize: '1.5vmin',
    backgroundColor:'white',
    borderRadius:'50px',
    width: '50%',
    height: '50%'
  }
  var style3 = {
    fontSize: '1.5vmin',
    boxSizing: 'border-box',
    padding: '20px'
  }
  var sh = {
    style: {
      color: 'red',
      backgroundColor: 'white',
      padding: '20px'
    }
  }
  var styleError = { backgroundColor: 'red', padding: '15px' }
  var props = {
    style: style3,
    width:'100%',
    height:'100%',
    frameBorder:'0',
    src: 'https://gitter.im/ESova/esova.github.io/~embed'
  }

  return function render404 () {
    db.get('error', function (error) {
      var vtree = h('div', { style: style1 }, [
        h('h1', sh, 'Something went wrong'),
        h('h2', sh, 'Please tell us about it in the chat so we can fix it :-)'),
        h('div', { style: style2 }, [
          h('iframe', props)
        ]),
        h('br'),
        h('span', 'Copy <ctrl-c> & paste <ctrl-v> the error to the chat'),
        h('xmp', { style: styleError, onclick: onclick }, [
          '```js\n',
          JSON.stringify(error, null, 4),
          '\n```'
        ])
      ])
      engine.push(vtree)
    })
  }
}
function onclick (event) {
  try {
    var range = document.createRange()
    range.selectNode(event.target)
    window.getSelection().addRange(range);
    document.execCommand('copy')
  }
  catch (err) { console.log('Oops, unable to copy') }
}
